File: app/src/main/java/ru/eugen/noteseugen/observe/Publisher.java
package ru.eugen.noteseugen.observe;

import java.util.ArrayList;
import java.util.List;

import ru.eugen.noteseugen.data.Card;

public class Publisher {
    private List<Observer> observers;

    public Publisher() {
        observers = new ArrayList<>();
    }
    public void subscribe(Observer observer) {
        observers.add(observer);
    }
    public void unsubscribe(Observer observer) {
        observers.remove(observer);
    }
    public void notifySingle(Card card) {
        for (Observer observer : observers) {
            observer.updateCard(card);
            unsubscribe(observer);
        }
    }
}

File: app/src/main/java/ru/eugen/noteseugen/observe/Publisher.kt
package ru.eugen.noteseugen.observe

import ru.eugen.noteseugen.data.Card

class Publisher {
    val observers: MutableList<Observer>

    constructor() {
        observers = ArrayList()
    }

    fun subscribe(observer:Observer):Unit {
        observers.add(observer)
    }

    fun unsubscribe(observer:Observer):Unit {
        observers.remove(observer)
    }

    fun notifySingle(card: Card):Unit {
        for (observer:Observer in observers) {
            observer.updateCard(card)
            unsubscribe(observer)
        }
    }
}
